src/xenia/apu/audio_system.cc

```diff
@@ -154,6 +171,156 @@ void AudioSystem::WorkerThreadMain() {
+
+        // Reset valid flags
+        data.input_buffer_0_valid = 0;
+        data.input_buffer_1_valid = 0;
+        data.output_buffer_valid = 0;
+
+        // Translate pointers for future use.
+        auto in0 = memory()->TranslatePhysical(data.input_buffer_0_ptr);
+        auto in1 = memory()->TranslatePhysical(data.input_buffer_1_ptr);
+        auto out = memory()->TranslatePhysical(data.output_buffer_ptr);
+
+        // I haven't seen this be used yet.
+        assert(!data.input_buffer_1_block_count);
+
+        // What I see:
+        // XMA outputs 2 bytes per sample
+        // 512 samples per frame (128 per subframe)
+        // Max output size is data.output_buffer_block_count * 256
+
+        // This decoder is fed packets (max 4095 per buffer)
+        // Packets contain "some" frames
+        // 32bit header (big endian)
+
+        // Frames are the smallest thing the SPUs can decode.
+        // They usually can span packets (libav handles this)
+
+        // Sample rates (data.sample_rate):
+        // 0 - 24 kHz ?
+        // 1 - 32 kHz
+        // 2 - 44.1 kHz ?
+        // 3 - 48 kHz ?
+
+        // SPUs also support stereo decoding. (data.is_stereo)
+
+        while (true) {
+          // Initial check - see if we've finished with the input
+          // TODO - Probably need to move this, I think it might skip the very
+          // last packet (see the call to PreparePacket)
+          size_t input_size = (data.input_buffer_0_block_count +
+                              data.input_buffer_1_block_count) * 2048;
+          size_t input_offset = (data.input_buffer_read_offset / 8 - 4);
+          size_t input_remaining = input_size - input_offset;
+          if (input_remaining == 0) {
+            // We're finished!
+            break;
+          }
+
+          // Now check the output buffer.
+          size_t output_size = data.output_buffer_block_count * 256;
+          size_t output_offset = data.output_buffer_write_offset * 256;
+          size_t output_remaining = output_size - output_offset;
+          if (output_remaining == 0) {
+            // Can't write any more data. Break.
+            // The game will kick us again with a new output buffer later.
+            break;
+          }
+
+          // This'll copy audio samples into the output buffer.
+          // The samples need to be 2 bytes long!
+          // Copies one frame at a time, so keep calling this until size == 0
+          int read = context.decoder->DecodePacket(out, output_offset,
+                                                   output_remaining);
+          if (read < 0) {
+            XELOGAPU("APU failed to decode packet (returned %.8X)", -read);
+            context.decoder->DiscardPacket();
+
+            // TODO: Set error state
+
+            break;
+          }
+
+          if (read == 0) {
+            // Select sample rate.
+            int sample_rate = 0;
+            if (data.sample_rate == 0) {
+              // TODO: Test this
```

Comment:
is it worth throwing asserts here so we can verify it works? or do you think they're pretty likely to be correct?